command.c

```diff
@@ -45,233 +46,74 @@
-    case CMD_REQ_REDIS_SINTER:
-    case CMD_REQ_REDIS_SINTERSTORE:
-    case CMD_REQ_REDIS_SREM:
-    case CMD_REQ_REDIS_SUNION:
-    case CMD_REQ_REDIS_SUNIONSTORE:
-    case CMD_REQ_REDIS_SRANDMEMBER:
-    case CMD_REQ_REDIS_SSCAN:
-
-    case CMD_REQ_REDIS_PFADD:
-    case CMD_REQ_REDIS_PFMERGE:
-    case CMD_REQ_REDIS_XACK:
-    case CMD_REQ_REDIS_XADD:
-    case CMD_REQ_REDIS_XAUTOCLAIM:
-    case CMD_REQ_REDIS_XCLAIM:
-    case CMD_REQ_REDIS_XDEL:
-    case CMD_REQ_REDIS_XPENDING:
-    case CMD_REQ_REDIS_XRANGE:
-    case CMD_REQ_REDIS_XREVRANGE:
-    case CMD_REQ_REDIS_XTRIM:
-    case CMD_REQ_REDIS_ZADD:
-    case CMD_REQ_REDIS_ZINTERSTORE:
-    case CMD_REQ_REDIS_ZRANGE:
-    case CMD_REQ_REDIS_ZRANGEBYSCORE:
-    case CMD_REQ_REDIS_ZREM:
-    case CMD_REQ_REDIS_ZREVRANGE:
-    case CMD_REQ_REDIS_ZRANGEBYLEX:
-    case CMD_REQ_REDIS_ZREVRANGEBYSCORE:
-    case CMD_REQ_REDIS_ZUNIONSTORE:
-    case CMD_REQ_REDIS_ZSCAN:
-        return 1;
+typedef enum {
+    KEYPOS_NONE,
+    KEYPOS_UNKNOWN,
+    KEYPOS_INDEX,
+    KEYPOS_KEYNUM
+} cmd_keypos;
+
+typedef struct {
+    cmd_type_t type;           /* A constant identifying the command. */
+    const char *name;          /* Command name */
+    const char *subname;       /* Subcommand name or NULL */
+    cmd_keypos firstkeymethod; /* First key none, unknown, pos or keynum */
+    int8_t firstkeypos;        /* Position of first key or the  arg */
+    int8_t arity;              /* Arity, neg number means min num args */
+} cmddef;
+
+/* Populate the table with code in cmddef.h generated from Redis JSON files. */
+static cmddef redis_commands[] = {
+#define COMMAND(_type, _name, _subname, _arity, _keymethod, _keypos)           \
+    {.type = CMD_REQ_REDIS_##_type,                                            \
+     .name = _name,                                                            \
+     .subname = _subname,                                                      \
+     .firstkeymethod = KEYPOS_##_keymethod,                                    \
+     .firstkeypos = _keypos,                                                   \
+     .arity = _arity},
+#include "cmddef.h"
+#undef COMMAND
+};
+
+/* Looks up a command or subcommand in the command table. Arg0 and arg1 are used
+ * to lookup the command. The function returns CMD_UNKNOWN on failure. On
+ * success, the command type is returned and *firstkey and *arity are
+ * populated. */
+cmddef *redis_lookup_cmd(const char *arg0, uint32_t arg0_len, const char *arg1,
+                         uint32_t arg1_len) {
+    int num_commands = sizeof(redis_commands) / sizeof(cmddef);
+    /* Find the command using binary search. */
+    int left = 0, right = num_commands - 1;
+    while (left <= right) {
+        int i = (left + right) / 2;
+        cmddef *c = &redis_commands[i];
+
+        int cmp = strncasecmp(c->name, arg0, arg0_len);
+        if (cmp == 0 && strlen(c->name) > arg0_len)
+            cmp = 1; /* "HGETALL" vs "HGET" */
```

Comment:
The lookup require a certain order of the commands, which should be correct since the python script does a sort on the filenames list (..I guess?).

I was just thinking that we sort the filenames, but we build the commands using the content of the files.
Maybe we can presume that the command-name and filename are the same, or maybe we should add a check/assert of it in the script to make sure.

Reply:
Good point. The python script should sort the list of commands instead. Or we can assert in (the python script) that the filename matches the command name within the file.